dkg/actions_active.go

```diff
@@ -0,0 +1,443 @@
+		return nil
+	}
+
+	// if there's an error sending to a party or saving the state, attempt a rollback by issuing an abort
+	rollback := func(err error) {
+		allParticipants := util.Concat(nextState.Joining, nextState.Remaining, nextState.Leaving)
+		d.log.Errorw("There was an error proposing a DKG", "err", err, "beaconID", beaconID)
+		_ = d.attemptAbort(ctx, me, allParticipants, beaconID)
+	}
+
+	return responseOrError(rollbackOnError(sendProposalToAllAndStoreState, rollback))
+}
+
+func (d *DKGProcess) StartAbort(ctx context.Context, options *drand.AbortOptions) (*drand.EmptyResponse, error) {
+	beaconID := options.BeaconID
+	d.log.Infow("Aborting DKG", "beaconID", beaconID)
+
+	me, err := d.identityForBeacon(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	current, err := d.store.GetCurrent(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	if !util.EqualParticipant(current.Leader, me) {
+		return nil, errors.New("cannot abort the DKG if you aren't the leader")
+	}
+
+	nextState, err := current.Aborted()
+	if err != nil {
+		return nil, err
+	}
+
+	allParticipants := util.Concat(nextState.Joining, nextState.Remaining, nextState.Leaving)
+	if err := d.attemptAbort(ctx, me, allParticipants, beaconID); err != nil {
+		return nil, err
+	}
+
+	if err := d.store.SaveCurrent(beaconID, nextState); err != nil {
+		d.log.Errorw("error aborting the DKG", "error", err, "beaconID", beaconID)
+	} else {
+		d.log.Infow("DKG aborted successfully", "beaconID", beaconID)
+	}
+	return responseOrError(err)
+}
+
+func (d *DKGProcess) StartExecute(ctx context.Context, options *drand.ExecutionOptions) (*drand.EmptyResponse, error) {
+	beaconID := options.BeaconID
+
+	stateTransition := func(me *drand.Participant, current *DBState) (*DBState, error) {
+		if !util.EqualParticipant(current.Leader, me) {
+			return nil, errors.New("cannot start execution if you aren't the leader")
+		}
+		return current.Executing(me)
+	}
+
+	callback := func(me *drand.Participant, nextState *DBState) error {
+		allParticipants := util.Concat(nextState.Joining, nextState.Remaining, nextState.Leaving)
+		return d.network.SendIgnoringConnectionError(
+			me,
+			allParticipants,
+			func(client net.DKGClient, peer net.Peer) (*drand.EmptyResponse, error) {
+				return client.Execute(ctx, peer, &drand.StartExecution{
+					Metadata: &drand.DKGMetadata{
+						BeaconID: beaconID,
+					}},
+				)
+			},
+		)
+	}
+
+	err := d.executeActionWithCallback("DKG execution", beaconID, stateTransition, callback)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// set up the DKG broadcaster for first so we're ready to broadcast DKG messages
+	dkgConfig, err := d.setupDKG(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	d.log.Infow("DKG execution leader setup successful", "beaconID", beaconID)
+
+	go func() {
+		// wait for `KickOffGracePeriod` to allow other nodes to set up their broadcasters
+		time.Sleep(d.config.KickoffGracePeriod)
```

Comment:
Is it not possible to use a fixed timestamp ? otherwise, the nodes that received the full proposal late will also start late because they'll also sleep this grace period.